lib/mint/core/headers.ex

```diff
@@ -0,0 +1,134 @@
+defmodule Mint.Core.Headers do
+  @type canonical_header() ::
+          {original_name :: String.t(), canonical_name :: String.t(), value :: String.t()}
+  @type raw_header() :: {original_name :: String.t(), value :: String.t()}
+
+  @unallowed_trailer_headers MapSet.new([
+                               "content-encoding",
+                               "content-length",
+                               "content-range",
+                               "content-type",
+                               "trailer",
+                               "transfer-encoding",
+
+                               # Control headers (https://svn.tools.ietf.org/svn/wg/httpbis/specs/rfc7231.html#rfc.section.5.1)
+                               "cache-control",
+                               "expect",
+                               "host",
+                               "max-forwards",
+                               "pragma",
+                               "range",
+                               "te",
+
+                               # Conditionals (https://svn.tools.ietf.org/svn/wg/httpbis/specs/rfc7231.html#rfc.section.5.2)
+                               "if-match",
+                               "if-none-match",
+                               "if-modified-since",
+                               "if-unmodified-since",
+                               "if-range",
+
+                               # Authentication/authorization (https://tools.ietf.org/html/rfc7235#section-5.3)
+                               "authorization",
+                               "proxy-authenticate",
+                               "proxy-authorization",
+                               "www-authenticate",
+
+                               # Cookie management (https://tools.ietf.org/html/rfc6265)
+                               "cookie",
+                               "set-cookie",
+
+                               # Control data (https://svn.tools.ietf.org/svn/wg/httpbis/specs/rfc7231.html#rfc.section.7.1)
+                               "age",
+                               "cache-control",
+                               "expires",
+                               "date",
+                               "location",
+                               "retry-after",
+                               "vary",
+                               "warning"
+                             ])
+
+  @spec from_raw_headers([raw_header()]) :: [canonical_header()]
+  def from_raw_headers(headers) do
+    Enum.map(headers, fn {name, value} -> {name, String.downcase(name, :ascii), value} end)
```

Comment:
Also, to make sure we don't fuck this up in the future:

```suggestion
    Enum.map(headers, fn {name, value} -> {name, lower_raw(name), value} end)
```

?